### PR TITLE
Add support for MSVC

### LINKED ISSUE #1372: Add support for MSVC
#1306 adds support for using setuptools, which knows how to invoke MSVC. However, initial experimentation revealed some follow-up work:

* Compiler flags need fixing (they are currently `-Werror` instead of `/W4` for instance)
* `.lib` files seem missing (https://travis-ci.org/themperek/cocotb/builds/646876123)
  * Can't use `dlltool`, need `lib /def:modelsim.def /out:modelsim.lib /machine:x86` etc

Opening this issue to track future work.

### PR DESCRIPTION
**Pull request by tpambor**
This PR adds support for compiling with MSVC.

Depends on:
- [x] #1795 Don't include unistd.h on windows
- [x] #1797 Remove unused includes in embed.h
- [x] #1796 Add __declspec(dllexport) for functions/classes/variables to be exported

TODO:
- [x] Add CI
- [x] Clarify what warning level is desired on MSVC (at the moment the default `/W3` is used), W4 per recommendation on https://docs.microsoft.com/de-de/cpp/build/reference/compiler-option-warning-level
- [x] #2334 Portable compiler/linker flags
- [x] Make Makefiles work without lib prefix

Closes #1372, #2334, #2341

### CODE REVIEW

**Comment by themperek** [comment]
If you change `py3_mingw` to `py3` CI should pick MSVC compiler.
https://github.com/cocotb/cocotb/blob/c7da01ce2d09aec7f76da9f1b802a3bd679942fc/.travis.yml#L88

Not sure if the right versions are available: https://docs.travis-ci.com/user/reference/windows/#pre-installed-chocolatey-packages

File: cocotb/share/lib/simulator/simulatormodule.cpp
Comment (eric-wieser) [inline cocotb/share/lib/simulator/simulatormodule.cpp:132]: I remember needing to forward-declare these in order to reference them below. Perhaps the references were removed in a later commit.

Does msvc reject these?

File: cocotb/share/lib/vpi/VpiTrampoline.cpp
Comment (eric-wieser) [inline cocotb/share/lib/vpi/VpiTrampoline.cpp:121]: You can avoid this special case with
```suggestion
        template<typename...Args>
        static PLI_INT32 vpi_vcontrol         (PLI_INT32 operation, Args... args)
```
Then use `f(operation, args...)` later.

File: cocotb/share/lib/simulator/simulatormodule.cpp
Comment (tpambor) [reply cocotb/share/lib/simulator/simulatormodule.cpp:132]:
```
  cocotb\share\lib\simulator\simulatormodule.cpp(964): error C2086: 'PyTypeObject `anonymous namespace'::gpi_hdl_Object<GpiObjHdl *>::py_type': redefinition
  cocotb\share\lib\simulator\simulatormodule.cpp(128): note: see declaration of 'py_type'
```
Without forward-decleration MSVC is happy but clang complains:
```
  cocotb/share/lib/simulator/simulatormodule.cpp:956:43: error: explicit specialization of 'py_type' after instantiation
  PyTypeObject gpi_hdl_Object<gpi_sim_hdl>::py_type = []() -> PyTypeObject {
                                            ^
  cocotb/share/lib/simulator/simulatormodule.cpp:485:59: note: implicit instantiation first required here
      if (Py_TYPE(pSigHdl) != &gpi_hdl_Object<gpi_sim_hdl>::py_type) {
```
GCC is fine either way

Comment (eric-wieser) [reply cocotb/share/lib/simulator/simulatormodule.cpp:132]: Any idea which compiler is following the standard here?

Comment (tpambor) [reply cocotb/share/lib/simulator/simulatormodule.cpp:132]: I think the relevant part from spec is §14.7.3 [temp.expl.spec]/p6 (emphasis added):
> If a template, a member template or a member of a class template is explicitly specialized then **that specialization shall be declared before the first use of that specialization that would cause an implicit instantiation to take place, in every translation unit in which such a use occurs**; no diagnostic is required.

I updated the code to move the template before its first use and now it works fine on all compilers, see [`078dc8b`](https://github.com/cocotb/cocotb/pull/1798/commits/078dc8b3cccae2fce9e8feaa1d862443a500c6fe).

Comment (eric-wieser) [reply cocotb/share/lib/simulator/simulatormodule.cpp:132]: I asked [on StackOverflow here](https://stackoverflow.com/q/61770519/102441). To me this looks like an MSVC bug, the existing code already contains the declaration.

Comment (am11) [reply cocotb/share/lib/simulator/simulatormodule.cpp:132]: @eric-wieser, I read your comment on community [bug report](https://developercommunity.visualstudio.com/content/problem/89837/vc2017-a-declaration-of-variable-template-is-alway.html) (which was mysteriously closed by some sort of a canned response..). However, [similar report](https://developercommunity.visualstudio.com/content/problem/319447/explicit-specialization-of-static-data-member-inco.html) was made where it was explained that it has been fixed in VS2019 16.5; but since it is a breaking change, it will require `/permissive-` to compile in standard conformance mode.

library-free repro:

```c++
// specialization-declaration.cpp
template<typename X> class A { };

template<typename Y> class B
{
public:
    static A<Y> j;
};

template<> A<int> B<int>::j;
A<int> y = B<int>::j;
template<> A<int> B<int>::j = { };
```
with:
```cmd
cl /c .\specialization-declaration.cpp
```
gives `C2086` errors, while in std conformance mode:

```cmd
cl /c /permissive- .\specialization-declaration.cpp
```
compiles fine:
```
Microsoft (R) C/C++ Optimizing Compiler Version 19.26.28806 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

specialization-declaration.cpp
```
(Visual Studio 2019, 16.6.3)

Comment (tpambor) [reply cocotb/share/lib/simulator/simulatormodule.cpp]: @am11 Thank you, adding `/permissive-` fixed the compile error.

**Comment by ktbarrett** [comment]
@tpambor Is #1783 *necessary* to enable MSVC support? It's a rather big change that some maintainers have issue with, and we will discuss it soon. It would be nice to just bring in the last 2 commits so we have MSVC support now, if at all possible.

I'd like these large changes in early in the release window so we have a few months to detect issues.

**Comment by tpambor** [comment]
@ktbarrett This could probably go in without #1783.

This PR is not yet ready to be merged. I have split out in #2019, #2020 the parts that are ready. These are cleanups not directly related with adding support for MSVC. Once they are merged I will give this PR a respin without #1783.

**Comment by ktbarrett** [comment]
@tpambor Is there any news on the updates? We are hoping to release 1.5 soon.

**Comment by tpambor** [comment]
At the moment this is blocked on #2334.

**Comment by ktbarrett** [comment]
Does your last commit in this PR not solve the problem?

**Comment by tpambor** [comment]
I did a clean respin without #1783. It already compiles fine and should not regress the other targets. Tests are currently still failing, which is probably due to problems with the manifest.

**Review by ktbarrett — COMMENTED** [review]
Looks pretty solid.

File: cocotb_build_libs.py
Comment (tpambor) [reply cocotb_build_libs.py:29]: C++11 should be the default.

> Note that there is no plan to add a C++11 switch.
https://devblogs.microsoft.com/cppblog/standards-version-switches-in-the-compiler/

File: cocotb_build_libs.py
Comment (eric-wieser) [inline cocotb_build_libs.py:387]: Why are we prefixing the libraries here? [this answer](https://stackoverflow.com/a/9796089/102441) seems to suggest the the library names should be unprefixed.

Comment (tpambor) [reply cocotb_build_libs.py:387]: If you pass `-lcocotb` to mingw it looks for `libcocotb.dll` while msvc looks for `cocotb.dll`. As we probably still want the library to be called `libcocotb.dll` we prefix it with lib if we compile with msvc.

File: cocotb_build_libs.py
Comment (tpambor) [reply cocotb_build_libs.py:366]: See https://github.com/cocotb/cocotb/pull/1798#discussion_r555233126_. I will add comment.

File: cocotb_build_libs.py
Comment (eric-wieser) [reply cocotb_build_libs.py:387]: I don't really agree with that. Under msvc, "libpython" is really called `python.dll`. I think we should try to stick with the platform defaults unless there is a good reason not to.

Presumably the issue is that the makefiles assume this prefix?

**Comment by tpambor** [comment]
> Where does the `lib` prefix get added?

It comes from the extension name `libcocotb` etc.

**Comment by eric-wieser** [comment]
> > Where does the `lib` prefix get added?
> 
> It comes from the extension name `libcocotb` etc.

I feel like we should not be putting `lib` in the extension name - so instead of removing the `lib` prefix on MSVC, we should be adding it on non-MSVC; the reason we need it is because we are abusing `Extension` to build shared libraries.

**Comment by themperek** [comment]
I see quite a persistence here 👍🏻 .
Good luck.

**Comment by themperek** [comment]
I think there was discussion software (cannot find it now). 
Do we really want to have a different name (no lib) just for msvc? 
Is this required?

File: cocotb_build_libs.py
Comment (themperek) [reply cocotb_build_libs.py]: In this case name of vpi/fli/vhpi library will depend on the compiler one uses? So depending on compiler mings/msvc different name has to be used?

File: cocotb/share/lib/gpi/GpiCommon.cpp
Comment (ktbarrett) [inline cocotb/share/lib/gpi/GpiCommon.cpp:190]: I was never a fan of this. I'm not sure why the user shouldn't just pass the full name. This prevents the ability to pass relative or absolute paths. I'll split that out into another issue.

File: cocotb_build_libs.py
Comment (themperek) [reply cocotb_build_libs.py]: Is not better to have a bit more "ugly" compilation and keep those names the same?

**Comment by ktbarrett** [comment]
@tpambor I realized that may change requires a lot of the makefile changes you have already done. So [here](https://github.com/ktbarrett/cocotb/tree/gpi-extra-libs) are my suggested changes to the `gpi_load_libs` code along with newsfragment and documentation change. Might save you a little work.

File: cocotb/share/makefiles/Makefile.inc
Comment (themperek) [inline cocotb/share/makefiles/Makefile.inc:71]: I do not have a very strong opinion here but is a good idea to have the name of vpi/fli/vhpi different only for one simulator.

**Comment by tpambor** [comment]
All tests are passing now 🚀

File: cocotb/share/makefiles/Makefile.inc
Comment (tpambor) [reply cocotb/share/makefiles/Makefile.inc:71]: That's not for a simulator here, mingw/msys uses the unix style `lib` prefix while microsoft convention on windows is without. At the moment this is mostly a change internal to cocotb but if `libgpi` or other components might be distributed separately, it may be a slight advantage to follow the native way.

Comment (themperek) [reply cocotb/share/makefiles/Makefile.inc:71]: Will it complicate potential extensions?

Comment (themperek) [reply cocotb/share/makefiles/Makefile.inc:71]: >That's not for a simulator here, mingw/msys uses the unix style lib prefix while microsoft convention on windows is without.

That is the point so now it is just for windows and msvc different. 

Is this really an advantage?

**Comment by tpambor** [comment]
> @tpambor I realized that my changes required a lot of the makefile changes you have already done. So [here](https://github.com/ktbarrett/cocotb/tree/gpi-extra-libs) are my suggested changes to the `gpi_load_libs` code along with newsfragment and documentation change. Might save you a little work.

I have merged your changes.

File: cocotb/share/lib/gpi/GpiCommon.cpp
Comment (ktbarrett) [inline cocotb/share/lib/gpi/GpiCommon.cpp:171]:
```suggestion
        auto const idx = arg.rfind(':');  // find from right since path could contain colons (Windows)
        if (idx == std::string::npos) {
```

File: cocotb/share/lib/gpi/GpiCommon.cpp
Comment (ktbarrett) [inline cocotb/share/lib/gpi/GpiCommon.cpp:178]:
```suggestion
        std::string const lib_name = arg.substr(0, idx);
        std::string const func_name = arg.substr(idx+1, std::string::npos);
```

**Review by ktbarrett — COMMENTED** [review]
Sorry, switched to `rfind` at the last minute without checking it. These changes work on your branch locally.

File: cocotb_build_libs.py
Comment (ktbarrett) [inline cocotb_build_libs.py:173]: I noticed it was creating a file on Linux.
```suggestion
        if os.name == "nt":
            create_sxs_appconfig(self.get_ext_fullpath(os.path.join("cocotb", "simulator")))
```

File: cocotb/share/makefiles/Makefile.inc
Comment (ktbarrett) [reply cocotb/share/makefiles/Makefile.inc:71]: Depending on the system that we use to load extensions, it might not be a problem. Since we are supporting full library names and paths in `GPI_EXTRA`, the file names in the list could be absolute paths to files on the system. These names would probably be populated with globing, so there is no issue of conditional computation of the library name.

I think our current use of `GPI_EXTRA` for loading "other" libraries for mixed language support will go the way of the dinosaur; and most of the library name computing added in this PR will be removed.

Comment (themperek) [reply cocotb/share/makefiles/Makefile.inc:71]: Maybe can add this to `cocotb-config` so custom flows is a bit easier to build?
Something like `cocotb-config --vpi-lib icarus`?

File: cocotb/share/lib/gpi/GpiCommon.cpp
Comment (tpambor) [reply cocotb/share/lib/gpi/GpiCommon.cpp]: Private CI still seems unhappy. I guess `GPI_EXTRA` needs to be changed in the Makefiles as well?

Comment (themperek) [reply cocotb/share/lib/gpi/GpiCommon.cpp]: Log [az.log](https://github.com/cocotb/cocotb/files/5814367/az.log)

Comment (ktbarrett) [reply cocotb/share/lib/gpi/GpiCommon.cpp]: Yeah my change requires full names. So .so/.dll extensions are necessary. The idea is whatever the user puts in GPI_EXTRA is fed directly to utils_dyn_open.

File: cocotb/share/makefiles/Makefile.inc
Comment (ktbarrett) [reply cocotb/share/makefiles/Makefile.inc:71]: I think so. `cocotb-config` should be a one stop shop for custom flows. Maybe we can do that as a follow on.

**Review by ktbarrett — COMMENTED** [review]
Can you also add a newsfragment for 1798 stating there is no MSVC support? So close...

File: .github/workflows/regression-tests.yml
Comment (tpambor) [reply .github/workflows/regression-tests.yml:261]: At least it helps track down problems with library loading on windows more quickly. Mostly it is helpful for PRs like #2315 which touch the dependencies between libraries. The normal error message doesn't provide a lot of details and with the additional output here you could get a hint what is wrong even if you don't have a windows setup.

**Comment by ktbarrett** [comment]
One last thing is to update the [installation docs](https://cocotb--1798.org.readthedocs.build/en/1798/install.html#installation) and the [development version installation docs](https://cocotb--1798.org.readthedocs.build/en/1798/install_devel.html#installing-the-development-version). For the typical end user who doesn't care which compiler is used, they don't need to specify the compiler and it should automatically pick the right one? 

Then I might link to the installation instructions in the newsfragment:
```rst
See :ref:`install` for more details. 
```

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
Going to get one more approval since this is a large change, I will give it one week.

**Comment by ktbarrett** [comment]
I will also test it locally and with Questa, just to make sure nothing got messed up on the systems we don't test in CI.

**Comment by ktbarrett** [comment]
Status of testing:

* It builds and runs on RHEL7 using Questa.
* It builds and runs on Windows using GCC in an Anaconda environment.
* It does not build in an MSYS2 environment using GCC (#2378), but I think we can skip that issue for now.
* It builds on Windows using MSVC in an Anaconda environment ___but does not run___.

I think your MSVC test is a red herring. Run `pip install cocotb` with the `-v` flag and see what compiler is being used. GCC is being installed previously to compile Icarus, GCC remains in the anaconda environment (see the log output when installing the runtime dependencies note they are already installed), GCC is picked up as the compiler and everything works.

**Comment by tpambor** [comment]
> * It does not build in an MSYS2 environment using GCC (#2378), but I think we can skip that issue for now.

Works for me locally with `pip install .`, note I am not using `pip -e` might be related to that.

> * It builds on Windows using MSVC in an Anaconda environment _**but does not run**_.

Works for me locally with
```
conda install -c msys2 m2-base m2-make
pip install .
make test
```
on a clean install of miniconda.

> I think your MSVC test is a red herring. Run `pip install cocotb` with the `-v` flag and see what compiler is being used. GCC is being installed previously to compile Icarus, GCC remains in the anaconda environment (see the log output when installing the runtime dependencies note they are already installed), GCC is picked up as the compiler and everything works.

Should be working, I will add `-v` again, this is also useful because I think LGTM only inspects the warnings for gcc on linux and therefore we also would have an impression for msvc and mingw gcc.

**Comment by ktbarrett** [comment]
Odd... it is using MSVC without any arguments despite GCC being installed.

I have an existing MSVC install. I did a fresh miniconda install. If I `pip install cocotb` it uses MSVC and builds successfully. As soon as I `conda install -c msys2 m2-base m2-make` and run `pip install .` it errors leaving me a traceback through `distutils/.../cygwincompiler.py`. So it seems to be trying to use GCC though I haven't installed it.

On your MSYS2 installation, when compiling, is it creating libraries with the `.so` extension or `.dll`? I get `.so`s which confuses `ld` when trying to link. I made this environment fresh, since I didn't have MSYS2 installed before yesterday.

**Comment by themperek** [comment]
>On your MSYS2 installation, when compiling, is it creating libraries with the .so extension or .dll? I get .sos which confuses ld when trying to link. I made this environment fresh since I didn't have MSYS2 installed before yesterday.

Straight MSYS2 never worked I believe. Need to use mingw64/32 https://github.com/cocotb/cocotb/wiki/Tier-2-Setup-Instructions#windows-installation-using-msys2.

**Comment by tpambor** [comment]
> Odd... it is using MSVC without any arguments despite GCC being installed.
> 
> I have an existing MSVC install. I did a fresh miniconda install. If I `pip install .` it uses MSVC and builds successfully. As soon as I `conda install -c msys2 m2-base m2-make` and run `pip install .` it errors leaving me a traceback through `distutils/.../cygwincompiler.py`. So it seems to be trying to use GCC though I haven't installed it.

For me it sticks to MSVC. I think normally the default is the compiler that was also used to compile python because I guess extension compiled with gcc would not work for python compiled with MSVC and viceversa.

```
(base) C:\>make
make: *** No targets specified and no makefile found.  Stop.      
(base) C:\>python
Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32
```


> On your MSYS2 installation, when compiling, is it creating libraries with the `.so` extension or `.dll`? I get `.so`s which confuses `ld` when trying to link. I made this environment fresh, since I didn't have MSYS2 installed before yesterday.

```
    Directory: C:\msys64\mingw64\lib\python3.8\site-packages\cocotb

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       22.01.2021     13:20                drivers
d-----       22.01.2021     13:20                generators
d-----       22.01.2021     13:20                libs
d-----       22.01.2021     13:20                monitors
d-----       22.01.2021     13:20                share
d-----       22.01.2021     13:20                vendor
d-----       22.01.2021     13:20                __pycache__
-a----       22.01.2021     13:20           3665 ANSI.py
-a----       22.01.2021     13:20          25333 binary.py
-a----       22.01.2021     13:20           8942 bus.py
-a----       22.01.2021     13:20           6468 clock.py
-a----       22.01.2021     13:20           6529 config.py
-a----       22.01.2021     13:20          21455 decorators.py
-a----       22.01.2021     13:20          36769 handle.py
-a----       22.01.2021     13:20           2926 ipython_support.py
-a----       22.01.2021     13:20          10365 log.py
-a----       22.01.2021     13:20           1913 memdebug.py
-a----       22.01.2021     13:20           1377 outcomes.py
-a----       22.01.2021     13:20          30200 regression.py
-a----       22.01.2021     13:20           5476 result.py
-a----       22.01.2021     13:20          35604 scheduler.py
-a----       22.01.2021     13:20          11336 scoreboard.py
-a----       22.01.2021     13:20          58880 simulator-cpython-38.dll
-a----       22.01.2021     13:20            263 simulator-cpython-38.dll.2.config
-a----       22.01.2021     13:20          30300 triggers.py
-a----       22.01.2021     13:20          21515 utils.py
-a----       22.01.2021     13:20           7036 wavedrom.py
-a----       22.01.2021     13:20           3630 xunit_reporter.py
-a----       22.01.2021     13:20           2752 _py_compat.py
-a----       22.01.2021     13:20           2928 _sim_versions.py
-a----       22.01.2021     13:20            308 _version.py
-a----       22.01.2021     13:20          11045 __init__.py

    Directory: C:\msys64\mingw64\lib\python3.8\site-packages\cocotb\libs

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       22.01.2021     13:20          19968 libcocotb.dll
-a----       22.01.2021     13:20          14848 libcocotbutils.dll
-a----       22.01.2021     13:20         101888 libcocotbvhpi_aldec.dll
-a----       22.01.2021     13:20          92160 libcocotbvpi_aldec.dll
-a----       22.01.2021     13:20          92160 libcocotbvpi_ghdl.dll
-a----       22.01.2021     13:20          92160 libcocotbvpi_icarus.vpl
-a----       22.01.2021     13:20          92160 libcocotbvpi_modelsim.dll
-a----       22.01.2021     13:20          15360 libembed.dll
-a----       22.01.2021     13:20          65536 libgpi.dll
-a----       22.01.2021     13:20          40448 libgpilog.dll
-a----       22.01.2021     13:20          40960 libpygpilog.dll
```

**Comment by ktbarrett** [comment]
@themperek Those instructions did not work for me. They are also out of date: you can no longer install from master.zip; and it mentions `pip3` and `python3`, but MSYS2 and MinGW32/64 have renamed `python` and `pip` for Python 3.

<details><summary>Output</summary>

```
kaleb@DESKTOP-2CHBMVI MINGW64 ~/cocotb
$ pip install .
Processing c:/msys64/home/kaleb/cocotb
    ERROR: Command errored out with exit status 1:
     command: C:/msys64/mingw64/bin/python.exe -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:/Users/kaleb/AppData/Local/Temp/pip-req-build-0uja_t1j/setup.py'"'"'; __file__='"'"'C:/Users/kaleb/AppData/Local/Temp/pip-req-build-0uja_t1j/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base C:/Users/kaleb/AppData/Local/Temp/pip-pip-egg-info-3y2f0v_u
         cwd: C:/Users/kaleb/AppData/Local/Temp/pip-req-build-0uja_t1j/
    Complete output (34 lines):
    Modelsim/Questa executable (vdbg) executable not found. No FLI interface will be available.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:/Users/kaleb/AppData/Local/Temp/pip-req-build-0uja_t1j/setup.py", line 84, in <module>
        setup(
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools/__init__.py", line 153, in setup
        return distutils.core.setup(**attrs)
      File "C:/msys64/mingw64/lib/python3.8/distutils/core.py", line 108, in setup
        _setup_distribution = dist = klass(attrs)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools/dist.py", line 430, in __init__
        _Distribution.__init__(self, {
      File "C:/msys64/mingw64/lib/python3.8/distutils/dist.py", line 292, in __init__
        self.finalize_options()
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools/dist.py", line 706, in finalize_options
        ep(self)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools/dist.py", line 713, in _finalize_setup_keywords
        ep.load()(self, ep.name, value)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/integration.py", line 26, in version_keyword
        dist.metadata.version = _get_version(config)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 173, in _get_version
        parsed_version = _do_parse(config)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 127, in _do_parse
        version = _version_from_entrypoints(config) or _version_from_entrypoints(
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 65, in _version_from_entrypoints
        version = _call_entrypoint_fn(root, config, ep.load())
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 46, in _call_entrypoint_fn
        return fn(root, config=config)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/git.py", line 97, in parse
        wd = GitWorkdir.from_potential_worktree(config.absolute_root)
      File "C:/msys64/mingw64/lib/python3.8/site-packages/setuptools_scm/git.py", line 33, in from_potential_worktree
        if not samefile(real_wd, wd):
      File "C:/msys64/mingw64/lib/python3.8/genericpath.py", line 100, in samefile
        s1 = os.stat(f1)
    FileNotFoundError: [WinError 3] The system cannot find the path specified: '/c/Users/kaleb/AppData/Local/Temp/pip-req-build-0uja_t1j'
    ----------------------------------------
WARNING: Discarding file:///C://msys64/home/kaleb/cocotb. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

</details>

@tpambor Apparently if I have the libpython conda package installed it tries to use GCC, but will use MSVC otherwise. Of course, if I don't have libpython installed it doesn't run correctly. I doesn't run correct even if I do install it after the fact. So the question is why does CI and your machine work when mine doesn't? Do you have an MSVC compiled libpythonXX.dll somewhere else on your system that it is picking up?

```
/c/iverilog/bin/iverilog -o sim_build/sim.vvp -D COCOTB_SIM=1 -s sample_module -f sim_build/cmds.f -g2012   /c/Users/kaleb/cocotb/tests/test_cases/issue_120/../../../tests/designs/sample_module/sample_module.sv
C:/iverilog/bin/iverilog.exe: error while loading shared libraries: ?: cannot open shared object file: No such file or directory
```

**Comment by marlonjames** [comment]
Tested, builds and runs:

Active-HDL 32-bit - Conda Python 3.8.5 (default, Sep  4 2020, 00:03:40) [MSC v.1916 32 bit (Intel)] :: Anaconda, Inc. on win32 - cocotb regression tests ✅ 
Active-HDL 32-bit - System Python 3.8.7 (tags/v3.8.7:6503f05, Dec 21 2020, 17:43:54) [MSC v.1928 32 bit (Intel)] on win32 - custom Tcl flow ✅

**Comment by ktbarrett** [comment]
@themperek Seems my issue is https://github.com/pypa/setuptools_scm/issues/415. Though that is not an issue this PR can address.

@tpambor I guess it was some issue with my Icarus. Even old versions work fine now. I think this PR is good to go.

File: documentation/source/newsfragments/1798.feature.rst
Comment (ktbarrett) [inline documentation/source/newsfragments/1798.feature.rst:2]: The issue with setuptools_scm means the mingw environments are not currently supported. That might be fixed upstream, or perhaps installing from sdists where setuptools_scm has already run would work? We don't currently have any instructions for how to use the official distribution for Python on Windows. I think we should remove this line.

**Comment by ktbarrett** [comment]
That is the last change I see necessary before merging. If you want to clean up the commits, now is the time. Alternatively, I can just squash them.

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
@tpambor Thanks for all the work, this is great! It was a long process, but I'm glad to finally see this come in. 

Waiting until Wednesday morning to see if there are any last minute comments, then I will rebase.

**Comment by ktbarrett** [comment]
Making sure the documentation build isn't an issue (it looks like it just blew up on their end).

**Comment by ktbarrett** [comment]
@tpambor Thanks again for the help! I don't think many of us here are very Windows savvy, so your contributions are greatly appreciated.

**Pull request merged**
